Code/Framework/AzCore/AzCore/IO/Streamer/StreamerComponent.cpp

```diff
@@ -208,7 +209,28 @@ namespace AZ
     {
         if (m_streamer)
         {
-            m_streamer->QueueRequest(m_streamer->Report(AZ::IO::Requests::ReportType::FileLocks));
+            auto locks = new AZStd::vector<AZ::IO::Statistic>();
+            AZ::IO::FileRequestPtr request = m_streamer->Report(*locks, AZ::IO::IStreamerTypes::ReportType::FileLocks);
+            auto callback = [locks](AZ::IO::FileRequestHandle)
+            {
+                if (!locks->empty())
+                {
+                    for (AZ::IO::Statistic& lock : *locks)
+                    {
+                        const AZStd::string* path = AZStd::get_if<AZStd::string>(&lock.GetValue());
+                        AZ_Printf(
+                            "Streamer", "File lock in %.*s : '%s'.\n", AZ_STRING_ARG(lock.GetName()),
+                            path ? path->c_str() : "<Unsupported path type>");
+                    }
+                }
+                else
+                {
+                    AZ_Printf("Streamer", "No files are locked.\n");
+                }
+                delete locks;
+            };
```

Comment:
I think encapsulating this using a unique_ptr would be better ti avoid the explicit delete call.

```suggestion
            auto locks = AZStd::make_unique<AZStd::vector<AZ::IO::Statistic>>();
            AZ::IO::FileRequestPtr request = m_streamer->Report(*locks, AZ::IO::IStreamerTypes::ReportType::FileLocks);
            auto callback = [statisticsLocks = AZStd::move(locks)](AZ::IO::FileRequestHandle)
            {
                AZStd::unique_ptr<AZStd::vector<AZ::IO::Statistic> locks = AZStd::move(statisticsLocks);
                if (!locks->empty())
                {
                    for (AZ::IO::Statistic& lock : *locks)
                    {
                        const AZStd::string* path = AZStd::get_if<AZStd::string>(&lock.GetValue());
                        AZ_Printf(
                            "Streamer", "File lock in %.*s : '%s'.\n", AZ_STRING_ARG(lock.GetName()),
                            path ? path->c_str() : "<Unsupported path type>");
                    }
                }
                else
                {
                    AZ_Printf("Streamer", "No files are locked.\n");
                }
            };
```

Reply:
There's a bug in the implementation of the small object optimization of `AZStd::function` I think because the above code results in a compiler error. In particular in `function_base.h(282,1)` where instead of moving the object it's copied. Perhaps an issue with move constructor of `AZStd::function`?  (Haven't looked deeply into it.)

Reply:
It is not actually a bug.
The std::function only supports [copy constructible and assignable](https://en.cppreference.com/w/cpp/utility/functional/function) function objects or pointers.

C++23 has addressed this with [the std::move_only_function](https://en.cppreference.com/w/cpp/utility/functional/move_only_function), but that isn't available to us.